File: days/src/main/java/aoc2024j/day10/Day10.java

package aoc2024j.day10;

import utils.IO;
import utils.IntGrid;

import java.util.HashSet;
import java.util.List;
import java.util.Set;
import java.util.function.Predicate;
import java.util.function.UnaryOperator;
import java.util.stream.IntStream;
import java.util.stream.Stream;

public class Day10 {

    enum Direction implements UnaryOperator<Position> {

        UP(0, -1), DOWN(0, 1), LEFT(-1, 0), RIGHT(1, 0);

        final int dx, dy;

        Direction(int dx, int dy) {
            this.dx = dx;
            this.dy = dy;
        }

        @Override
        public Position apply(Position position) {
            return new Position(position.x() + dx, position.y() + dy);
        }

        static Stream<Position> expand(Position position) {
            return Stream.of(Direction.values())
                    .map(d -> d.apply(position));
        }
    }

    record Position(int x, int y) {
    }

    record Path(Position head, Path tail) {

        static Path of(Position position) {
            return new Path(position, null);
        }

        Path add(Position position) {
            return new Path(position, this);
        }
    }

    static class TopographicalMap extends IntGrid implements Predicate<Position> {
        public TopographicalMap(List<String> data) {
            super(data);
        }

        int get(Position p) {
            return points[p.y()][p.x()];
        }

        @Override
        public boolean test(Position position) {
            return position.y() >= 0 && position.y() < height &&
                    position.x() >= 0 && position.x() < width;
        }

        Stream<Position> trailHeads() {
            return IntStream.range(0, height)
                    .boxed()
                    .flatMap(y ->
                            IntStream.range(0, width)
                                    .mapToObj(x -> new Position(x, y)))
                    .filter(p -> get(p) == 0);
        }

        Stream<Position> expand(Position position) {
            return Direction.expand(position)
                    .filter(this);
        }

        Set<Path> trailsFrom(Position trailHead) {
            var trails = new HashSet<Path>();
            trails(Path.of(trailHead), trails);
            return trails;
        }

        void trails(Path path, Set<Path> found) {
            var position = path.head();
            int value = get(position);
            if (value == 9) {
                found.add(path);
            } else {
                expand(position)
                        .filter(p -> get(p) == value + 1)
                        .forEach(p -> trails(path.add(p), found));
            }
        }

        long part1() {
            return trailHeads()
                    .flatMap(p -> trailsFrom(p).stream().map(Path::head).distinct())
                    .count();
        }

        long part2() {
            return trailHeads()
                    .map(this::trailsFrom)
                    .mapToLong(Set::size)
                    .sum();
        }
    }

    long part1(List<String> data) {
        return new TopographicalMap(data).part1(); //totalScore1();
    }

    long part2(List<String> data) {
        return new TopographicalMap(data).part2(); //totalScore2();
    }

    public static void main() {
        var day10 = new Day10();
        var data = IO.getResourceAsList("aoc2024/day10.txt");
        var part1 = day10.part1(data);
        System.out.println("part1 = " + part1);
        var part2 = day10.part2(data);
        System.out.println("part2 = " + part2);
    }
}

File: days/src/main/scala/aoc2024s/day10/Day10.scala

package aoc2024s.day10;

import utils.IO;
import scala.jdk.CollectionConverters.*

object Day10 {

  enum Direction(val dx: Int, val dy: Int) {
    case Up extends Direction(0, -1)
    case Right extends Direction(1, 0)
    case Down extends Direction(0, 1)
    case Left extends Direction(-1, 0)

    def apply(p: Position): Position =
      Position(p.x + dx, p.y + dy)
  }

  object Direction {
    def expand(p: Position): Seq[Position] =
      Direction.values.map(_.apply(p))
  }

  case class Position(x: Int, y: Int)

  type Path = List[Position]

  case class TopographicalMap(values: Array[Array[Int]]) {
    val height: Int = values.length
    val width: Int = values(0).length

    def apply(p: Position): Int =
      values(p.y)(p.x)

    def isInside(p: Position): Boolean =
      p.x >= 0 && p.x < width && p.y >= 0 && p.y < height

    def this(data: List[String]) =
      this(data.map(_.map {
        _ - '0'
      }.toArray).toArray)

    def trailheads: Seq[Position] =
      for
        y <- 0 until height
        x <- 0 until width
        if values(y)(x) == 0
      yield Position(x, y)

    def expand(p: Position): Seq[Position] =
      Direction.expand(p).filter(isInside)

    def trails(start: Position): Seq[Path] =
      trails(List(start))

    def trails(path: Path): Seq[Path] = {
      val at = path.head
      val value = this (at)
      if value == 9 then Seq(path)
      else {
        expand(at)
          .filter(p => this (p) == value + 1)
          .flatMap(p => trails(p :: path))
      }
    }

    def part1: Long =
      trailheads
        .flatMap(trails(_).map(_.head).distinct)
        .size

    def part2: Long =
      trailheads
        .flatMap(trails)
        .size
  }

  def part1(data: List[String]): Long = {
    new TopographicalMap(data).part1
  }

  def part2(data: List[String]): Long = {
    new TopographicalMap(data).part2
  }

  @main def main10(): Unit = {
    val data = IO.getResourceAsList("aoc2024/day10.txt").asScala.toList;
    val part1 = Day10.part1(data);
    println(s"part1 = $part1");
    val part2 = Day10.part2(data);
    println(s"part2 = $part2");
  }
}
